### ISSUE TITLE
Exact APC hit path (0.6.17): single-row warm cache merged into Batch* caches, Qwen3.5 shortcut copies the whole cache per decode token (6.5 -> 9.9 tok/s at 33k; patch)

### ISSUE THREAD

**Issue by felix-ab** [Blaizzy/mlx-vlm#2210]
## Exact APC hit path: a single-row warm cache is merged into Batch* caches, and Qwen3.5's singleton shortcut then copies the whole KV/SSM cache on every decode token (6.5 → 9.9 tok/s at 33k)

_Measured on 0.6.17. 0.7.0 (released 2026-09-07) redesigned APC and still has `make_warm_batch_exact_cache_multi`; I have not re-measured the hit path on 0.7.0, so this is filed as a 0.6.x report and a data point for the redesign._

mlx-vlm 0.6.17, mlx 0.32.2, Mac mini M4 Pro 48 GB, `orcarouter/Qwen3.8-27B-Uncensored-MLX` 6-bit (hybrid GatedDeltaNet + attention), server with `APC_ENABLED=1`, exact cache, one active sequence.

**Symptom.** Decode on an exact-cache hit at 30–33k tokens ran at 6.4–6.7 tok/s while a cold (non-hit) decode at the same context ran at 9.9 tok/s; the footprint was also 36 GB vs 28 GB.

**Cause.** `make_warm_batch_exact_cache_multi` merges the stored row(s) into batch-aware caches even when there is exactly one row. Qwen3_5's language model takes its single-row batch-cache shortcut for those, which extracts and re-merges (copies) the entire KV + recurrent state on every decode step — O(context) work per token.

**Fix that worked** (25 lines, env-gated `APC_EXACT_SINGLE_ROW_PLAIN`, default on): when `len(row_caches) == 1` and no kv quantization is configured, return the plain single-row clone (evaluated) instead of merging, clearing a zero left-padding marker on `ArraysCache` entries. Output-identical (seeded before/after runs); decode at 33k 6.5 → 9.8–10.1 tok/s, footprint 36 → 28 GB; in production since 2026-09-04. Patch: `mlx-vlm-0.6.17-apc.patch` in https://gist.github.com/felix-ab/78ab16b7d15ba32432e9d8d972cd5ca9. The function still exists on main with the same merge-first structure, but I have only measured 0.6.17.

**Comment by gtonic** [comment]
Re-measured on **0.7.0**, which you flagged as the open question. It reproduces — but only with speculative decoding **off**, and that turned out to be the interesting part.

## Setup

- Apple M5 Pro, 48 GB, macOS 26, `iogpu.wired_limit_mb=40960` (40 GiB working set)
- mlx-vlm 0.7.0, mlx + mlx-metal 0.32.2, mlx-lm 0.31.3
- `mlx-community/Qwen3.8-27B-4bit` (`qwen3_5`, hybrid GDN + full attention, 64 layers, 16 full-attention) — the 4-bit affine conversion, not your 6-bit one
- Server, continuous batching, **one** sequence, exact APC with disk tier, f16 KV (no `--kv-bits`)
- 28,590-token prompt, 300 decoded tokens, `temperature 0`, `caffeinate -dimsu`

Method: each pair sends the same prompt twice. The cold arm carries a nonce prefix so nothing can match; the warm arm is the identical prompt. Context is read as `prompt_n + cache_n`, because on a hit `prompt_n` is only the unmatched remainder (1 token here). A pair whose arms differ in context, or whose cold arm hits, is discarded rather than averaged.

## Four arms

| # | speculative decoding | local patches in the APC path | cold tok/s | warm tok/s | warm/cold |
|---|---|---|---:|---:|---:|
| A | DFlash 2, `block_size 4` | applied | 20.44 / 20.11 / 20.80 | 20.69 / 20.07 / 20.87 | **1.003** |
| B | DFlash 2, `block_size 4` | reverted | 20.73 / 20.90 | 20.81 / 21.01 | **1.004** |
| C | off | applied | 15.96 / 15.73 | 10.28 / 10.16 | **0.645** |
| D | off | reverted | 15.73 / 15.50 | 9.58 / 10.03 | **0.628** |

Arms B and D revert the two local patches of ours that touch this code path at all (one suppresses the redundant second exact snapshot per request, one carries #2072's ownership transfer). They change nothing about the result, which is why I am comfortable reporting it as upstream behaviour: **−35% to −39% on a warm hit, without a drafter, on 0.7.0.**

With a drafter the effect is gone — 1.003 and 1.004 across five pairs, i.e. inside the run-to-run spread. If the per-decode-token copy were merely amortised over a verification block, at `block_size 4` and 47% acceptance (~2.9 accepted tokens per target pass) a 35% per-token penalty should still show up as roughly −15%. It does not show up at all, which suggests the speculative path does not take that shortcut rather than paying it less often. I have not read far enough into it to state that as fact.

## The structural half of your report is still true on 0.7.0

`make_warm_batch_exact_cache_multi` (`mlx_vlm/apc.py:3949`) still merges unconditionally — there is no `len(row_caches) == 1` shortcut — and it is reached from `apc_coordinator.py:119` and `generate/ar.py:2700`. What I could not locate on 0.7.0 is the second half: `extract_prompt_cache_from_batch` has exactly one caller in the release, and it sits inside `apc.py` rather than in a model decode path.

## Footprint

Directionally consistent with your 36 vs 28 GiB, though noisier than the throughput. Peak `sum` observed inside the request window, per arm: the cold arms sat at 23.1–25.2 GiB, and two of the four warm arms spiked to **33.8** and **34.9 GiB** against a 40 GiB working set, while the other two stayed at 21.5 and 23.4. On a machine that runs at 95% of its working set that spread is its own problem, independent of the tok/s.

## Why this may have gone unnoticed

Every APC number I had before today was a *prefill* number (89,630 ms → 350 ms on a warm hit at 36k). A warm hit that decodes at two thirds of the cold rate still looks like an unambiguous win in that metric, because the prefill it saves is worth minutes per turn and the decode it costs is worth seconds — until the answer is long. And with a drafter on, which is the default for this model here, the decode penalty is not there at all.

Happy to run 33k, other block sizes, or an `--kv-bits` arm if any of those would help pin the path down.

**Comment by althere-am** [comment]
# RFC: Avoid per-token cache repacking in Qwen single-row batched decoding while preserving correctness and fast prefill

Adding an independent reproduction and a proposed design/validation plan to this existing issue. Our measurements below concern the older installed mlx-vlm 0.6.4 environment; they are not a claim that we tested upstream 0.7.0. Unlike an APC-hit-only comparison, the main reproduction uses an in-process BatchGenerator without persistent prefix caching.

## Summary

We found a reproducible performance issue in the Qwen single-row batched execution path in `mlx-vlm`.

When a batch contains one sequence, `Qwen3_5Model.__call__` extracts per-row caches, runs the model on those caches, and merges them back into batch caches. Repeating this operation for every generated token introduces context-dependent overhead and substantial temporary memory allocations.

Disabling this branch in an isolated benchmark restores decoding throughput close to the non-batched library path. However, unconditional disabling can slow prefill by changing attention-mask handling and preventing oMLX’s FA-256 kernel from being selected.

We propose a guarded fast path that avoids repeated cache repacking while preserving padding, position, recurrent-state, and prefix-cache semantics.

## Tested environment

- Apple Mac mini, M4 Pro, 48 GB unified memory.
- `mlx` 0.32.0.
- `mlx-lm` 0.31.3.
- `mlx-vlm` 0.6.4 as installed in the oMLX environment.
- `oMLX` 0.6.4.
- Models:
  - `mlx-community/Qwen3.6-35B-A3B-4bit`.
  - `mlx-community/Qwen3.8-27B-4bit`.
- One active request, thinking disabled, greedy decoding.
- Non-quantized KV cache.
- Prefill chunk size: 2048 tokens.
- Target input lengths: 4K, 16K, 32K; actual counts within ±3%.

The issue reproduces in an in-process `mlx_lm.BatchGenerator` benchmark without the oMLX scheduler, HTTP transport, or persistent prefix cache.

## Observed behavior

For Qwen3.6-35B-A3B:

| Execution path | Decode 4K / 16K / 32K, tokens/s | Process memory footprint |
|---|---:|---:|
| Non-batched generation, oMLX model-loading path and patches | 83.6 / 71.7 / 62.2 | 22–23 GiB |
| BatchGenerator, existing single-row branch | 51.5 / 40.6 / 30.5 | 24–36 GiB |
| BatchGenerator, branch disabled, patches retained | 82.3 / 71.3 / 61.9 | 22–24 GiB |
| BatchGenerator, branch retained, all 12 tested kernel patches disabled | 52.5 / 40.8 / 30.6 | Up to 35 GiB |
| BatchGenerator, branch and tested kernel patches disabled | 80.8 / 69.7 / 61.1 | 22–25 GiB |

The initial comparison used 768 output tokens and three repetitions per input length. Branch-ablation runs used 256 output tokens and two repetitions; a matching control with the branch enabled measured 53.2 / 41.1 / 30.7 tokens/s.

The effect also reproduces on Qwen3.8-27B:

| Execution path | Decode 4K / 16K / 32K, tokens/s |
|---|---:|
| Non-batched library path | 14.7 / 13.6 / 12.8 |
| BatchGenerator, branch enabled | 12.7 / 10.15 / 7.95 |
| BatchGenerator, branch disabled | 14.65 / 13.5 / 12.35 |

These Qwen3.8 runs used 256 output tokens and two repetitions per condition.

## Suspected mechanism

In `mlx_vlm/models/qwen3_5/language.py`, the single-row batch branch performs:

```text
For each layer cache:
    extract row 0

Run the model using the extracted caches

For each layer cache:
    merge the updated row back into a batch cache
```

Relevant operations in `mlx_lm/models/cache.py` include:

- `BatchKVCache.extract`: scalar padding access through `.item()` and contiguous KV slices.
- Updating the extracted `KVCache`: potentially growing a buffer that has no spare capacity.
- `BatchKVCache.merge`: allocation and population of new batch buffers.
- Corresponding extraction/merge operations for GDN recurrent states.

This creates repeated work proportional to accumulated context length. Exact physical copying depends on array layout and MLX execution optimizations; we do not assume every operation always materializes a separate full copy.

A synthetic cache-only microbenchmark measured 3.65 / 8.18 / 14.25 ms per extract-update-merge step at 4K / 16K / 32K. This accounts for approximately 50–86% of the measured decoding gap. Attribution of the remainder is not yet established.

## Why unconditional disabling is insufficient

With oMLX patches retained, disabling the branch increases Qwen3.6 prefill TTFT:

- 16K: 19.15 → 22.91 seconds.
- 32K: 46.83 → 62.18 seconds.

The oMLX FA-256 routing predicate accepts `None` or `"causal"` masks, but rejects array masks. The ordinary batch path can supply an array mask, preventing this optimized attention path from being selected.

This explains a routing difference, but not every observed prefill interaction. With both the branch and all tested kernel patches disabled, 32K TTFT was 49.26 seconds.

For short outputs, slower prefill can outweigh faster decoding.

## Proposed approaches

### A. Preserve a persistent single-sequence cache representation

Use the existing prefill path, then retain the single-sequence caches throughout decoding.

Convert back to batch representation only when required, such as:

- Another sequence joins the batch.
- The scheduler extracts or persists cache state.
- A prefix-cache or GDN snapshot is created.
- The request completes.

This requires auditing every consumer that expects batch-cache representation.

### B. Support a guarded fast path directly on BatchKVCache

Avoid extract/merge while operating directly on the batch cache.

Select a simpler attention mask only when semantic equivalence is established, including:

- Batch size one.
- No relevant padding.
- Consistent logical positions and cache offsets.
- No attention to unused buffer capacity.
- Correct handling of restored prefixes.

Retain the existing path for unsupported cases.

Neither approach should introduce unconditional per-token host synchronization to check these conditions.

## Validation requirements

Correctness should precede performance acceptance:

- Compare logits and generated tokens against the existing reference path; investigate differences beyond expected numerical variation.
- Test padded and unpadded inputs.
- Test chunked prefill and chunk boundaries.
- Test batch transitions, especially 2 → 1 and 1 → 2.
- Test prefix reuse, changed facts, and GDN-state restoration.
- Test text and multimodal position handling.
- Verify structured output and tool calling in the serving integration.
- Keep quantized caches and speculative decoding explicitly out of scope unless separately validated.

Performance acceptance should demonstrate:

- Decoding close to the non-batched reference.
- No material prefill regression.
- Elimination of most excess temporary memory.
- Sustained behavior across consecutive requests in the actual server.

## Scope and questions for maintainers

This is a demonstrated performance issue in the tested execution path. We have not established the introducing commit, so we are not yet calling it a version regression.

The branch-disabled benchmarks used synthetic text and counting output. They establish performance behavior, not production readiness or full semantic equivalence.

Questions:

1. Which correctness cases motivated the current single-row fallback?
2. Should the fix live in mlx-vlm’s model path, mlx-lm’s cache API, or their integration?
3. Is a persistent single-row view feasible without repeated copying?
4. Would maintainers prefer a minimal correctness/performance reproducer before an implementation PR?

Related implementation: LM Studio’s MLX Engine already narrows `_is_single_row_batch_cache` handling based on padding and cache type. It may be useful prior art, but we have not validated it as a drop-in fix for oMLX:

https://github.com/lmstudio-ai/mlx-engine/blob/main/mlx_engine/model_kit/patches/qwen3_5.py
